src/coprocessor/dag/aggr_fn/impl_bit_op.rs

```diff
@@ -0,0 +1,460 @@
+use cop_datatype::builder::FieldTypeBuilder;
+use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeFlag, FieldTypeTp};
+use tipb::expression::{Expr, ExprType, FieldType};
+
+use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::codec::mysql::Tz;
+use crate::coprocessor::dag::expr::EvalContext;
+use crate::coprocessor::dag::rpn_expr::{RpnExpression, RpnExpressionBuilder};
+use crate::coprocessor::{Error, Result};
+
+/// A trait for all bit operations
+pub trait BitOp: Clone + std::fmt::Debug + Send + Sync + 'static {
+    /// Returns the bit operation type
+    fn tp() -> ExprType;
+
+    /// Returns the bit operation initial state
+    fn init_state() -> usize;
+
+    /// Executes the special bit operation
+    fn op(lhs: &mut usize, rhs: usize);
+}
+
+macro_rules! bit_op {
+    ($name:ident, $tp:path, $init:tt, $op:tt) => {
+        #[derive(Debug, Clone, Copy)]
+        pub struct $name;
+        impl BitOp for $name {
+            fn tp() -> ExprType {
+                $tp
+            }
+
+            fn init_state() -> usize {
+                $init
+            }
+
+            fn op(lhs: &mut usize, rhs: usize) {
+                *lhs $op rhs
+            }
+        }
+    };
+}
+
+bit_op!(BitAnd, ExprType::Agg_BitAnd, 0xffffffffffffffff, &=);
+bit_op!(BitOr, ExprType::Agg_BitOr, 0, |=);
+bit_op!(BitXor, ExprType::Agg_BitXor, 0, ^=);
+
+/// The parser for bit operation aggregate functions.
+pub struct AggrFnDefinitionParserBitOp<T: BitOp>(pub std::marker::PhantomData<T>);
+
+impl<T: BitOp> super::AggrDefinitionParser for AggrFnDefinitionParserBitOp<T> {
+    fn check_supported(&self, aggr_def: &Expr) -> Result<()> {
+        assert_eq!(aggr_def.get_tp(), T::tp());
+
+        // Check whether or not the children's field type is supported. Currently we only support
+        // Int and does not support other types (which need casting).
+        // TODO: remove this check after implementing `CAST as Int`
+        let child = &aggr_def.get_children()[0];
+        let eval_type = EvalType::try_from(child.get_field_type().tp())
+            .map_err(|e| Error::Other(box_err!(e)))?;
+        match eval_type {
+            EvalType::Int => {}
+            _ => return Err(box_err!("Cast from {:?} is not supported", eval_type)),
+        }
+
+        super::util::check_aggr_exp_supported_one_child(aggr_def)
+    }
+
+    fn parse(
+        &self,
+        mut aggr_def: Expr,
+        time_zone: &Tz,
+        // We use the same structure for all data types, so this parameter is not needed.
+        src_schema: &[FieldType],
+        out_schema: &mut Vec<FieldType>,
+        out_exp: &mut Vec<RpnExpression>,
+    ) -> Result<Box<dyn super::AggrFunction>> {
+        assert_eq!(aggr_def.get_tp(), T::tp());
+
+        // bit operation outputs one column.
+        out_schema.push(
+            FieldTypeBuilder::new()
+                .tp(FieldTypeTp::LongLong)
+                .flag(FieldTypeFlag::UNSIGNED)
```

Comment:
I guess this schema can be built from `aggr_def.field_type`

Reply:
The output field type always uint64. https://github.com/pingcap/tidb/blob/e9653675dd3b996b47647fde7e623afaaaf4317c/executor/aggfuncs/func_bitfuncs.go

Reply:
I mean you can directly use aggr_def.field_type instead of build a new one